be/dtm0_log.c

```diff
@@ -0,0 +1,658 @@
+		break;
+	case M0_DTML_REDO:
+	default:
+		M0_IMPOSSIBLE("");
+	}
+}
+
+M0_INTERNAL int m0_be_dtm0_log_create(struct m0_be_tx        *tx,
+                                      struct m0_be_seg       *seg,
+                                      struct m0_be_dtm0_log **out)
+{
+	struct m0_be_dtm0_log *log;
+
+	M0_PRE(tx != NULL);
+	M0_PRE(seg != NULL);
+
+	M0_BE_ALLOC_PTR_SYNC(log, seg, tx);
+	M0_ASSERT(log != NULL);
+
+	M0_BE_ALLOC_PTR_SYNC(log->u.dl_list, seg, tx);
+	M0_ASSERT(log->u.dl_list != NULL);
+
+	log->dl_seg = seg;
+
+	lrec_be_list_create(log->u.dl_list, tx);
+	M0_BE_TX_CAPTURE_PTR(seg, tx, log);
+	*out = log;
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_destroy(struct m0_be_tx        *tx,
+                                        struct m0_be_dtm0_log **log)
+{
+#if 0 /* TODO: write down persistent implementation later */
+	struct m0_be_dtm0_log *plog = *log;
+
+	M0_PRE(plog != NULL);
+	m0_free(plog->u.dl_tlist);
+	m0_free(plog);
+	*log = NULL;
+#endif
+}
+
+M0_INTERNAL
+struct m0_dtm0_log_rec *m0_be_dtm0_log_find(struct m0_be_dtm0_log    *log,
+                                            const struct m0_dtm0_tid *id)
+{
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	if (log->dl_is_persistent) {
+		struct m0_dtm0_log_rec *lrec;
+
+		m0_be_list_for(lrec, log->u.dl_list, lrec) {
+			if (m0_dtm0_tid_cmp(log->dl_cs, &lrec->dlr_txd.dtd_id,
+                                          id) == M0_DTS_EQ) {
+				break;
+			}
+		} m0_be_list_endfor;
+		return lrec;
+	} else {
+		return m0_tl_find(lrec, rec, log->u.dl_tlist,
+			          m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id,
+                                          id) == M0_DTS_EQ);
+	}
+}
+
+static int be_dtm0_log_rec_init(struct m0_dtm0_log_rec **rec,
+				struct m0_be_tx         *tx,
+				struct m0_dtm0_tx_desc  *txd,
+				struct m0_buf           *payload)
+{
+	int                     rc;
+	struct m0_dtm0_log_rec *lrec;
+
+	M0_ALLOC_PTR(lrec);
+	if (lrec == NULL)
+		return M0_ERR(-ENOMEM);
+
+	rc = m0_dtm0_tx_desc_copy(txd, &lrec->dlr_txd);
+	if (rc != 0) {
+		m0_free(lrec);
+		return rc;
+	}
+
+	rc = m0_buf_copy(&lrec->dlr_payload, payload);
+	if (rc != 0) {
+		m0_dtm0_tx_desc_fini(&lrec->dlr_txd);
+		m0_free(lrec);
+		return rc;
+	}
+
+	*rec = lrec;
+	return 0;
+}
+
+static int be_dtm0_plog_rec_init(struct m0_dtm0_log_rec **out,
+				 struct m0_be_tx         *tx,
+				 struct m0_be_seg        *seg,
+				 struct m0_dtm0_tx_desc  *txd,
+				 struct m0_buf           *payload)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	M0_BE_ALLOC_PTR_SYNC(rec, seg, tx);
+	M0_ASSERT(rec != NULL);
+
+	rec->dlr_txd.dtd_id = txd->dtd_id;
+	rec->dlr_txd.dtd_ps.dtp_nr = txd->dtd_ps.dtp_nr;
+	M0_BE_ALLOC_ARR_SYNC(rec->dlr_txd.dtd_ps.dtp_pa,
+			     txd->dtd_ps.dtp_nr, seg, tx);
+	M0_ASSERT(rec->dlr_txd.dtd_ps.dtp_pa != NULL);
+
+	memcpy(rec->dlr_txd.dtd_ps.dtp_pa, txd->dtd_ps.dtp_pa,
```

Comment:
Why don't we call m0_dtm0_tx_desc_copy()?

Reply:
m0_dtm0_tx_desc_copy() performs an M0_ALLOC_ARR internally in m0_dtm0_tx_desc_init(). So, can't use here.